0000-deb-repo-additional-headers.md

```diff
@@ -0,0 +1,222 @@
+         +- Packages.xz
+         . . .
+      +- binary-i386
+      . . .
+   +- universe
+   . . .
+   +- Release                     <- Plain file
+   +- InRelease                   <- Release file + inline signature
+   +- Release.gpg                 <- Separate signature
+
+```
+
+E.g. for official Ubuntu repos the URL of these files:
+```
+http://mirror.example.com/ubuntu/dists/bionic-updates/InRelease
+http://mirror.example.com/ubuntu/dists/bionic-updates/Release.gpg
+```
+
+There's also an alternative layout for Debian repos called "flat repos". In this case the repo directory containins both `Release*` and `Packages*` files. E.g.:
+```
+https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
+```
+
+_Note #1:_
+
+In Uyuni the URL of a deb repository must point to the directory containing the binary packages for a specific architecture. E.g. for Ubuntu 18.04 updates repo is something like:
+```
+http://mirror.example.com/ubuntu/dists/bionic-updates/main/binary-amd64/
+```
+or in case of a flat repository to the directory containing both `Release*` and `Packages*` files:
+```
+https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
+```
+
+_Note #2:_
+
+Debian repos don't have a standard location for the public GPG keys used to verify the metadata. Sometimes the key is located in the root of a flat repo but the location is not standardized.
+
+### GPG keys management
+
+Uyuni ships the SUSE and OpenSUSE keys in the package `uyuni-build-keys`. The keys for Ubuntu and Debian should be shipped in this package too.
+
+Any additional keys that need to be used (e.g. for Ubuntu derivatives distros or for custom repos) must be created manually using the existing UI (`Systems -> Autoinstallation -> GPG and SSL keys`)
+
+
+### GPG Key usage in the UI
+
+The UI for creating/updating a repository (`Software -> Manage -> Repositories`) must be extended with one additional field:
+- GPG key
+
+This field must be visible only if the selected repository type is `deb` and the checkbox `Has Signed Metadata` is checked. For `deb` repositories the `Has Signed Metadata` checkbox must be enabled by default.
+
+The field must be shown as a selector. The values used to populate the selector come from the existing table `rhnCryptoKey`. Only keys for current organization must be shown.
+
+An additional option `Default keys` must shown in the selector as default. If this option is selected, the validation will be done using the GPG keys from package `uyuni-build-keys`.
+
+When saving the repository data, in case there was a GPG key selected other then `Default keys` the assignment will be saved into the new table `rhnContentSourceGpg`.
+
+### Key usage in the API
+
+The API method `channel.software.create_repo()` must be extended with one optional arguments to allow specifying the `GPG key ID`.
```

Comment:
Do we have an API method to get a list of the GPG keys?

Reply:
Yes, `kickstart.keys.listAllKeys`